minecraft-common/src/main/java/com/sigmundgranaas/forgero/minecraft/common/toolhandler/VeinMiningStrategy.java

```diff
@@ -2,96 +2,111 @@
-        //Reversing the direction to make the algorithm prioritize blocks in other direction that up and down
-        Collections.reverse(directions);
-        for (Direction direction : directions) {
-            BlockPos newBlock = pos.offset(direction, 1);
-
-            if (
-                    !blockSet.contains(newBlock)
-            ) {
-                if (world.getBlockState(newBlock).getBlock() == rootBlock) {
-                    depth -= 1;
-                    blockSet.add(newBlock);
-                    queue.add(newBlock);
-                }
-
-                for (Direction direction2 : directions) {
-                    BlockPos newBlock2 = newBlock.offset(direction2, 1);
-                    if (world.getBlockState(newBlock2).getBlock() == rootBlock && !blockSet.contains(newBlock2)) {
-                        depth -= 1;
-                        blockSet.add(newBlock2);
-                        queue.add(newBlock2);
-                    }
-                }
-                if (depth < 1) {
-                    return;
-                }
-            }
-        }
-        calculateNextBlocks(blockSet, queue, depth, rootBlock, world, player);
-    }
+	private final VeinBreaking handler;
 
+	private int currentDepth;
+
+	public VeinMiningStrategy(VeinBreaking handler) {
+		this.handler = handler;
+	}
+
+	public VeinMiningStrategy(PropertyData data) {
+		this.handler = new VeinBreaking((int) data.getValue(), data.getTags().get(0), data.getDescription());
+	}
+
+	@Override
+	public Set<BlockPos> getAvailableBlocks(BlockView world, BlockPos rootPos, PlayerEntity player) {
+		BlockState rootState = world.getBlockState(rootPos);
+		this.currentDepth = handler.depth();
+
+		if (BlockBreakingHandler.isBreakableBlock(world, rootPos, player) && rootState.isIn(TagKey.of(Registry.BLOCK_KEY, new Identifier(handler.tag())))) {
+			return getVeinMineableBlocksAtPosition(rootPos, world);
+		}
+
+		return Collections.emptySet();
+	}
+
+	private HashSet<BlockPos> getVeinMineableBlocksAtPosition(BlockPos rootBlockPos, BlockView world) {
+		var rootBlockState = world.getBlockState(rootBlockPos);
+
+		// Take a block
+		// Check around it in every direction
+		// For each direction also check around it in every direction
+		// Continue until depth <= 0
+		// Then reset depth to handler.depth() and go to 2 (the next direction)
+
+		var veinMineableBlocks = new HashSet<BlockPos>();
+		veinMineableBlocks.add(rootBlockPos);
+		var blocksToScan = new HashSet<BlockPos>();
+		blocksToScan.add(rootBlockPos);
+		var scannedBlocks = new HashSet<BlockPos>();
+
+		for (int i = this.currentDepth; i > 0 && !blocksToScan.isEmpty(); i--) {
+			HashSet<BlockPos> newBlocksToScan = new HashSet<>();
+
+			for (BlockPos blockToScanPos : blocksToScan) {
+				if (scannedBlocks.contains(blockToScanPos)) {
+					continue;
+				}
+
+				var blocksAroundScannedBlock = getBlocksAroundBlock(blockToScanPos);
+				blocksAroundScannedBlock.forEach(blockPos -> {
```

Comment:
Would adding a filter checking if these blocks are already in the scanned set before the forEach operation reduce some overhead from adding the same block to newBlocksToScan?

`blocksAroundScannedBlock.stream(block -> !scannedBlocks.isIn(block)).forEach...`

Reply:
That is such a small optimisation, I think it'll barely matter.
I don't really feel like testing it at the moment though. We can have a look at further optimisations later in a separate PR, if you don't mind.

Really happy this is currently working as intended without any visible lag spikes already.